arangod/Graph/Enumerators/WeightedTwoSidedEnumerator.cpp

```diff
@@ -621,50 +658,49 @@ void WeightedTwoSidedEnumerator<QueueType, PathStoreType, ProviderType,
+    } else {
+      auto searchLocation = getBallToContinueSearch();
+      if (searchLocation == BallSearchLocation::LEFT) {
+        _left.computeNeighbourhoodOfNextVertex(_right, _candidatesStore);
+      } else if (searchLocation == BallSearchLocation::RIGHT) {
+        _right.computeNeighbourhoodOfNextVertex(_left, _candidatesStore);
+      } else {
+        TRI_ASSERT(searchLocation == BallSearchLocation::FINISH);
+        // Our queue is empty. We cannot produce more results.
+        setAlgorithmFinished();
+      }
     }
 
-    if (matchPathLength > 0) {
-      // means we've found a match (-1.0 means no match)
-      if (_bestCandidateLength < 0 || matchPathLength < _bestCandidateLength) {
-        _bestCandidateLength = matchPathLength;
-      } else if (matchPathLength > _bestCandidateLength) {
-        // If a candidate has been found, we insert it into the store and only
-        // return the length of that path. As soon as we insert in into the
-        // store we sort that internal vector, therefore it might re-balance. If
-        // we pop the first element it is guaranteed that the calculated weight
-        // must be the same. It is not guaranteed that this is the path we
-        // actually inserted. Nevertheless, this should not be any problem.
-        //
-        // Additionally, this is only a potential candidate. We may find same
-        // paths multiple times. If that is the case, we just get rid of that
-        // candidate and do not pass it to the result data structure.
-        //
-        // Not only one candidate needs to be verified, but all candidates
-        // which do have lower weights as the current found (match).
-
-        while (std::get<0>(_candidatesStore.peek()) < matchPathLength) {
+    if (!_candidatesStore.isEmpty()) {
+      auto bestWeight = std::get<0>(_candidatesStore.peek());
+      // if the sum of the diameters of left and right search are bigger
+      // than the best candidate, there will not be a better candidate found.
+      //
+      // A simple shortest path search is done *now* (and not earlier!);
+      //
+      // It is *required* to continue search for a shortest path even
+      // after having found *some* path between the two searches:
+      // There might be improvements on the weight in paths that are
+      // found later. Improvements are impossible only if the sum of
+      // the diameters of the two searches is smaller than the current
+      // best found path.
+      //
+      // For a K-SHORTEST-PATH search all candidates that have lower weight
+      // than the sum of the two diameters are valid shortest paths that must
+      // be returned.
+      // K-SHORTEST-PATH search has to continue until the queues on both
+      // sides are empty
+
+      auto leftDiameter = _left.getDiameter();
+      auto rightDiameter = _right.getDiameter();
+      auto sumDiameter = leftDiameter + rightDiameter;
+
+      if (sumDiameter >= bestWeight) {
```

Comment:
Is this safe w.r.t. floating-point inaccuracies?

Reply:
I am trying to understand what the potential problems are that could pop up.

If you are referring to the `>=` part of this inequality, I don't think it will make a difference to put a `>` there; sure, one can construct cases in which one then searches a really long time to terminate instead of getting an answer quickly but you'd have to put effort in to do it. 

Further,

* There cannot be (obvious) catastrophic cancellation, as we sum up and compare non-negative numbers 
* Weights along paths are added up and as such, in particular if the magnitude of those weights varies, there'll be loss of precision. This is in particular also a problem for the diameter. There are ways of addressing this, none of which would fit into this PR.
* The diameters are of different magnitude and the diameter of one of the balls is rounded away; This would indeed make sumDiameter smaller than it should be this only makes a difference if this "rounding error" leads to early termination and potentially an incorrect result. 

I will have to do a slightly more careful numerical analysis to see whether there are any more pitfalls; There is certainly a small pandora's box that could be opened with respect to floating points and weighted path search;

In all, though, this will not introduce worse behaviour than before!